File: 42-trapping-rain-water/42-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(h) {
    let l = new Array(h.length);
    let r = new Array(h.length);
    let n = h.length-1;
    l[0] = h[0];
    r[n] = h[n];
    
    for(let i=1;i<=n;i++){
        l[i] = Math.max(l[i-1],h[i]);
    }
    for(let i=n-1;i>=0;i--){
        r[i] = Math.max(r[i+1],h[i]);
    }
    let ans = 0;
    for(let i=0;i<=n;i++){
        ans += Math.min(l[i],r[i]) - h[i];
    }
    return ans;
};

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        vector<int> l(h.size());l[0] = h[0];
        vector<int> r(h.size());r.back() = h.back();
        for(int i=1;i<h.size();i++){
            l[i] = max(l[i-1],h[i]);
        }
        for(int i=h.size()-2;i>=0;i--){
            r[i] = max(r[i+1],h[i]);
        }
        
        int ans = 0;
        for(int i=0;i<h.size();i++){
            ans += min(l[i],r[i]) - h[i];
        }
        return ans;
    }
};